## -------------------- ~ Revenue Leakage Detection & Profit Optimization - Amazon ~ --------------------

## Project Overview: 
This project builds a complete, production-ready analytics pipeline on an Amazon sales dataset of 100,000 orders spanning January 2020 to December 2024. The goal is to identify every source of revenue leakage — discounting anomalies, low-margin categories, high-loss regions, suspicious pricing inconsistencies — and then derive actionable profit optimization recommendations.
### The pipeline is organized into four phases:
- Phase 1 — Data Cleaning & Validation in Python (Pandas / NumPy)
- Phase 2 — Business Insights & Revenue Leakage Detection in SQL (MySQL)
- Phase 3 — Profit Optimization Analysis in SQL (advanced joins, aggregations)
- Phase 4 — Interactive Dashboard Design in Power BI

### *A] PHASE 1*
#### Data Cleaning & Validation: 
Before any analysis can be trusted, the raw data must be audited, cleaned, and feature-engineered.

### *Step 1:  Environment Setup & Library Imports*
We start every notebook by importing all libraries and suppressing noisy warnings.
NumPy and Pandas handle data manipulation;
Matplotlib and Seaborn provide visualization for EDA

In [1]:
# ── Cell 1: Library Imports ──
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
from tabulate import tabulate

pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: f'{x:,.2f}')
print('Libraries loaded successfully!')
print(f'Pandas  : {pd.__version__}')
print(f'NumPy   : {np.__version__}')

Libraries loaded successfully!
Pandas  : 2.3.3
NumPy   : 2.3.5


#### *✦ Conclusion:*
All required libraries imported. pandas.set_option ensures floats are readable in the notebook — important for revenue figures in the hundreds of thousands.

### *Step 2: Load Dataset & Initial Inspection*
We load the CSV and immediately run shape, dtypes, head(), and describe() to understand what we're working with before touching anything.

In [2]:
# Load data
df = pd.read_csv('C:/Users/HOLE/DATA ANALYST/Python Projects/Revenue Leakage Detection & Profit Optimization/Amazon.csv')


def print_heading(title):
    print("\n" + "="*50)
    print(title.center(50))
    print("="*50)

# Dataset overview
print_heading("DATASET OVERVIEW")
print(f"Shape (Rows, Columns): {df.shape}")

# Column data types
print_heading("COLUMN DATA TYPES")
print(df.dtypes)

# First 5 rows (leave it as is)
print_heading("FIRST 5 ROWS")
display(df.head())   # use display instead of print

# Statistical summary
print_heading("STATISTICAL SUMMARY")
display(df.describe())


                 DATASET OVERVIEW                 
Shape (Rows, Columns): (100000, 20)

                COLUMN DATA TYPES                 
OrderID           object
OrderDate         object
CustomerID        object
CustomerName      object
ProductID         object
ProductName       object
Category          object
Brand             object
Quantity           int64
UnitPrice        float64
Discount         float64
Tax              float64
ShippingCost     float64
TotalAmount      float64
PaymentMethod     object
OrderStatus       object
City              object
State             object
Country           object
SellerID          object
dtype: object

                   FIRST 5 ROWS                   


,OrderID,OrderDate,CustomerID,CustomerName,ProductID,ProductName,Category,Brand,Quantity,UnitPrice,Discount,Tax,ShippingCost,TotalAmount,PaymentMethod,OrderStatus,City,State,Country,SellerID
0,ORD0000001,2023-01-31,CUST001504,Vihaan Sharma,P00014,Drone Mini,Books,BrightLux,3,106.59,0.00,0.00,0.09,319.86,Debit Card,Delivered,Washington,DC,India,SELL01967
1,ORD0000002,2023-12-30,CUST000178,Pooja Kumar,P00040,Microphone,Home & Kitchen,UrbanStyle,1,251.37,0.05,19.10,1.74,259.64,Amazon Pay,Delivered,Fort Worth,TX,United States,SELL01298
2,ORD0000003,2022-05-10,CUST047516,Sneha Singh,P00044,Power Bank 20000mAh,Clothing,UrbanStyle,3,35.03,0.10,7.57,5.91,108.06,Debit Card,Delivered,Austin,TX,United States,SELL00908
3,ORD0000004,2023-07-18,CUST030059,Vihaan Reddy,P00041,Webcam Full HD,Home & Kitchen,Zenith,5,33.58,0.15,11.42,5.53,159.66,Cash on Delivery,Delivered,Charlotte,NC,India,SELL01164
4,ORD0000005,2023-02-04,CUST048677,Aditya Kapoor,P00029,T-Shirt,Clothing,KiddoFun,2,515.64,0.25,38.67,9.23,821.36,Credit Card,Cancelled,San Antonio,TX,Canada,SELL01411



               STATISTICAL SUMMARY                


,Quantity,UnitPrice,Discount,Tax,ShippingCost,TotalAmount
count,"100,000.00","100,000.00","100,000.00","100,000.00","100,000.00","100,000.00"
mean,3.00,302.91,0.07,68.47,7.41,918.26
std,1.41,171.84,0.08,74.13,4.32,724.51
min,1.00,5.00,0.00,0.00,0.00,4.27
25%,2.00,154.19,0.00,15.92,3.68,340.89
50%,3.00,303.07,0.05,45.25,7.30,714.32
75%,4.00,451.50,0.10,96.06,11.15,"1,349.76"
max,5.00,599.99,0.30,538.46,15.00,"3,534.98"


#### *✦ Conclusion*
The dataset has 100,000 rows and 20 columns. order_date is stored as a string object — it must be converted to datetime.
All numeric columns are correctly typed. No structural problems detected.

### *Step 3: Null Values & Duplicate Check*

In [3]:
# ── Cell 3: Null & Duplicate Check ──
print('=== NULL VALUES ===')
print(df.isnull().sum())
print('\nTotal nulls:', df.isnull().sum().sum())

print('\n=== DUPLICATE ROWS ===')
dup_count = df.duplicated().sum()
print(f'Duplicate rows: {dup_count}')

print('\n=== ORDER_ID UNIQUENESS ===')
print(f'Unique order_ids: {df["OrderID"].nunique()} / {len(df)}')

=== NULL VALUES ===
OrderID          0
OrderDate        0
CustomerID       0
CustomerName     0
ProductID        0
ProductName      0
Category         0
Brand            0
Quantity         0
UnitPrice        0
Discount         0
Tax              0
ShippingCost     0
TotalAmount      0
PaymentMethod    0
OrderStatus      0
City             0
State            0
Country          0
SellerID         0
dtype: int64

Total nulls: 0

=== DUPLICATE ROWS ===
Duplicate rows: 0

=== ORDER_ID UNIQUENESS ===
Unique order_ids: 100000 / 100000


#### *✦ Conclusion*
Zero nulls, zero duplicates, and OrderID is perfectly unique. The dataset is structurally clean. We proceed to logical validation.

#### Step 4: Fix Data Types — Parse OrderDate
String dates cannot be used for time-series analysis. We convert OrderDate to datetime and extract useful temporal features: year, month, quarter, and day-of-week.
These become critical columns for trend analysis in Phase 2.

In [4]:
# 1. Convert aur sirf lowercase features create karein
df['OrderDate'] = pd.to_datetime(df['OrderDate'])
df['year'] = df['OrderDate'].dt.year
df['month'] = df['OrderDate'].dt.month
df['quarter'] = df['OrderDate'].dt.quarter
df['month_name'] = df['OrderDate'].dt.month_name()
df['day_of_week'] = df['OrderDate'].dt.day_name()

# 2. Summary Info
years = sorted(df['year'].unique())
print(f"Date range: {df['OrderDate'].min().date()} to {df['OrderDate'].max().date()}")
print(f"Years covered: {years}\n")

# 3. Sample display ke liye sirf lowercase columns select karein
sample = df[['OrderDate', 'year', 'month', 'quarter', 'day_of_week']].head().copy()

# 4. Format for Readability (Sirf display ke liye names thode saaf kiye hain)
sample['OrderDate'] = sample['OrderDate'].dt.date
sample.columns = ['Order Date', 'Year', 'Month', 'Quarter', 'Day of Week']

print(tabulate(sample, headers='keys', tablefmt='simple', showindex=False))

Date range: 2020-01-01 to 2024-12-29
Years covered: [np.int32(2020), np.int32(2021), np.int32(2022), np.int32(2023), np.int32(2024)]

Order Date      Year    Month    Quarter  Day of Week
------------  ------  -------  ---------  -------------
2023-01-31      2023        1          1  Tuesday
2023-12-30      2023       12          4  Saturday
2022-05-10      2022        5          2  Tuesday
2023-07-18      2023        7          3  Tuesday
2023-02-04      2023        2          1  Saturday


#### *✦ Conclusion*
OrderDate is now a proper datetime object. Five new temporal columns are added.
These will power monthly trend and seasonality analysis in SQL and Power BI.

In [5]:
df.columns

Index(['OrderID', 'OrderDate', 'CustomerID', 'CustomerName', 'ProductID',
       'ProductName', 'Category', 'Brand', 'Quantity', 'UnitPrice', 'Discount',
       'Tax', 'ShippingCost', 'TotalAmount', 'PaymentMethod', 'OrderStatus',
       'City', 'State', 'Country', 'SellerID', 'year', 'month', 'quarter',
       'month_name', 'day_of_week'],
      dtype='object')

In [6]:
# Un columns ki list jo humein nahi chahiye
redundant_cols = ['Month Name', 'Day Of Week', 'Year', 'Month', 'Quarter']

# Inhe drop karein (errors='ignore' taaki agar koi missing ho toh error na aaye)
df = df.drop(columns=redundant_cols, errors='ignore')

# Ab check karein, sirf lowercase bache honge
print("Current Columns:", df.columns.tolist())

Current Columns: ['OrderID', 'OrderDate', 'CustomerID', 'CustomerName', 'ProductID', 'ProductName', 'Category', 'Brand', 'Quantity', 'UnitPrice', 'Discount', 'Tax', 'ShippingCost', 'TotalAmount', 'PaymentMethod', 'OrderStatus', 'City', 'State', 'Country', 'SellerID', 'year', 'month', 'quarter', 'month_name', 'day_of_week']


In [7]:
print(df.columns)

Index(['OrderID', 'OrderDate', 'CustomerID', 'CustomerName', 'ProductID',
       'ProductName', 'Category', 'Brand', 'Quantity', 'UnitPrice', 'Discount',
       'Tax', 'ShippingCost', 'TotalAmount', 'PaymentMethod', 'OrderStatus',
       'City', 'State', 'Country', 'SellerID', 'year', 'month', 'quarter',
       'month_name', 'day_of_week'],
      dtype='object')


In [8]:
# 1. Identify duplicates by looking at lowercase versions of names
cols_to_keep = ~df.columns.str.lower().duplicated()

# 2. Filter the DataFrame
df = df.loc[:, cols_to_keep]

# 3. Verify the result
print(df.columns.tolist())

['OrderID', 'OrderDate', 'CustomerID', 'CustomerName', 'ProductID', 'ProductName', 'Category', 'Brand', 'Quantity', 'UnitPrice', 'Discount', 'Tax', 'ShippingCost', 'TotalAmount', 'PaymentMethod', 'OrderStatus', 'City', 'State', 'Country', 'SellerID', 'year', 'month', 'quarter', 'month_name', 'day_of_week']


#### *✦ Conclusion*

### *Step 5: Logical Validation — Pricing Consistency*
The dataset contains UnitPrice, Quantity, Discount, Tax, ShippingCost, and TotalAmount. A key revenue leakage detection step is verifying the calculation.
Any mismatch indicates a pricing error, data entry mistake, or possible fraud.

In [9]:
# ── Cell 5: Validate TotalAmount Calculation ──
df['calc_total'] = np.round(
(df['UnitPrice'] * df['Quantity'] * (1 - df['Discount'])) + df['Tax'] + df['ShippingCost'], 2
)
df['revenue_error'] = np.abs(df['TotalAmount'] - df['calc_total']) > 0.05
errors = df[df['revenue_error']]
print(f'Rows with total_amount mismatch: {len(errors)}')
print('\nAll pricing formulas are consistent!')

Rows with total_amount mismatch: 0

All pricing formulas are consistent!


#### *✦ Conclusion*
Pricing logic is mathematically consistent across all 100,000 rows. No phantom discounts or miscalculated revenues detected at the raw data level.

### *Step 6  Outlier Detection with NumPy — IQR Method*
Outliers in price, TotalAmount, and Quantity can distort aggregated analytics.
We use the IQR (Interquartile Range) method — the industry standard — to flag statistical extremes.

In [10]:
# ── Cell 6: IQR-Based Outlier Detection ──
def detect_outliers(series, col_name):
    Q1  = np.percentile(series, 25)
    Q3  = np.percentile(series, 75)
    IQR = Q3 - Q1
    
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    
    mask = (series < lower) | (series > upper)
    count = mask.sum()
    pct   = (count / len(series)) * 100

    print(f"{col_name}")
    print(f"Outliers : {count} ({pct:.2f}%)")
    print(f"Range    : [{lower:.2f}, {upper:.2f}]")
    print("-" * 40)

    return mask


print("\nOUTLIER REPORT")
print("=" * 40)

for col in ['UnitPrice', 'Discount', 'Quantity', 'TotalAmount']:
    detect_outliers(df[col], col)


OUTLIER REPORT
UnitPrice
Outliers : 0 (0.00%)
Range    : [-291.78, 897.47]
----------------------------------------
Discount
Outliers : 1989 (1.99%)
Range    : [-0.15, 0.25]
----------------------------------------
Quantity
Outliers : 0 (0.00%)
Range    : [-1.00, 7.00]
----------------------------------------
TotalAmount
Outliers : 1360 (1.36%)
Range    : [-1172.42, 2863.08]
----------------------------------------


#### *✦ Conclusion*
The dataset contains 1,989 outliers in Discount and 1,360 outliers in TotalAmount. Because the percentage of outliers is very small (~1-2%), and because excessive discounts and extreme high-value orders are precisely the focus of our revenue leakage analysis, we will leave them in the dataset for SQL to dissect.

### *Step 7: Feature Engineering — Revenue Leakage Indicators*
This is the most critical step for revenue leakage detection. We engineer new columns that will serve as the foundation for all SQL analysis and Power BI visuals.

In [11]:
# Feature Engineering
df['max_possible_revenue'] = np.round(df['UnitPrice'] * df['Quantity'], 2)
df['revenue_lost_to_discount'] = np.round(df['max_possible_revenue'] * df['Discount'], 2)
df['revenue_efficiency'] = np.round(df['TotalAmount'] / df['max_possible_revenue'], 4)
df['is_high_discount'] = (df['Discount'] > 0.20).astype(int)
df['is_lost_order'] = df['OrderStatus'].isin(['Cancelled', 'Returned']).astype(int)

# Summary values
summary = [
    ["Total Max Possible Revenue", f"${df['max_possible_revenue'].sum():,.2f}"],
    ["Revenue Lost to Discounts", f"${df['revenue_lost_to_discount'].sum():,.2f}"],
    ["Overall Revenue Efficiency", f"{df['revenue_efficiency'].mean()*100:.2f}%"],
    ["High Discount Orders (>20%)", f"{df['is_high_discount'].sum():,}"],
    ["Lost Orders (Cancelled/Returned)", f"{df['is_lost_order'].sum():,}"]
]

# Print table
print("\nNEW FEATURE SUMMARY\n")
print(tabulate(summary, headers=["Metric", "Value"], tablefmt="simple"))


NEW FEATURE SUMMARY

Metric                            Value
--------------------------------  --------------
Total Max Possible Revenue        $91,021,212.99
Revenue Lost to Discounts         $6,783,123.94
Overall Revenue Efficiency        102.37%
High Discount Orders (>20%)       6,975
Lost Orders (Cancelled/Returned)  6,077


#### *✦ Conclusion*
This is the headline finding of Phase 1: $6.78 million in potential revenue is lost strictly to discounts. However, unlike typical leakage scenarios, the overall revenue efficiency is 102.37%. This means the additional revenue collected from shipping costs and taxes completely offsets the discount leakage at a macro level. The pricing baseline is actually strong. Therefore, the primary leakage hotspots are now isolated to the 6,975 high-discount orders and the 6,077 cancelled or returned orders. These specific cohorts — rather than the entire dataset — will be the primary targets for SQL optimization in Phase 2.

#### *Step 8: Statistical Summary by Category*
To transition to SQL, we first generate a high-level Python summary to see how that $6.78M in leakage is distributed across categories.

In [12]:
# ── Cell 8: Grouped Summary Statistics ───
cat_summary = df.groupby('Category').agg(
    total_orders   = ('OrderID', 'count'),
    avg_price      = ('UnitPrice', 'mean'),
    avg_discount   = ('Discount', 'mean'),
    total_revenue  = ('TotalAmount', 'sum'),
    revenue_lost   = ('revenue_lost_to_discount', 'sum')
).round(2).sort_values('total_revenue', ascending=False)

# Format columns (this is the key step)
cat_summary['total_revenue'] = cat_summary['total_revenue'].map('{:,.2f}'.format)
cat_summary['revenue_lost']  = cat_summary['revenue_lost'].map('{:,.2f}'.format)
cat_summary['avg_price']     = cat_summary['avg_price'].map('{:,.2f}'.format)
cat_summary['avg_discount']  = cat_summary['avg_discount'].map('{:.2f}'.format)

print("\nCATEGORY SUMMARY\n")

print(tabulate(
    cat_summary.reset_index(),
    headers='keys',
    tablefmt='simple',
    showindex=False
))


CATEGORY SUMMARY

Category             total_orders    avg_price    avg_discount  total_revenue    revenue_lost
-----------------  --------------  -----------  --------------  ---------------  --------------
Electronics                 16853       302.93            0.07  15,584,217.18    1,144,068.84
Sports & Outdoors           16804       301.76            0.07  15,345,571.88    1,132,324.12
Books                       16752       302.02            0.07  15,261,837.01    1,132,720.45
Clothing                    16439       305.2             0.07  15,253,397.50    1,127,571.05
Toys & Games                16542       302.81            0.07  15,216,684.99    1,117,830.41
Home & Kitchen              16610       302.76            0.08  15,163,939.36    1,128,609.07


#### *✦ Conclusion*
Electronics leads in total revenue (dollar 5.58M) and also has the highest revenue lost (dollar 1.14M). Critically, your data reveals a nearly identical average discount of exactly 7% across almost every category (with Home & Kitchen slightly higher at 8%). This confirms a rigid, blanket discount policy. Instead of strategic, category-specific pricing, the business is applying a flat ~7% discount everywhere, leading to over $6.7M in uniform leakage.

#### *Step 9: Export Cleaned Dataset for SQL Import*

In [13]:
# ── Cell 9: Export Cleaned Data ──────────────────────────────
export_cols = [
    'OrderID', 'OrderDate', 'CustomerID', 'ProductID', 'Category', 'Brand',
    'UnitPrice', 'Quantity', 'Discount', 'Tax', 'ShippingCost', 'TotalAmount',
    'PaymentMethod', 'OrderStatus', 'Country', 'SellerID',
    'max_possible_revenue', 'revenue_lost_to_discount', 'revenue_efficiency',
    'is_high_discount', 'is_lost_order', 'year', 'month', 'quarter', 'month_name'
]
df_clean = df[export_cols].copy()
df_clean['OrderDate'] = df_clean['OrderDate'].dt.strftime('%Y-%m-%d')
df_clean.to_csv('amazon_clean.csv', index=False)
print('Exported: amazon_clean.csv')

Exported: amazon_clean.csv


#### *✦ Conclusion*
Phase 1 Complete. A clean, 100,000-row dataset is ready for MySQL import. All leakage indicators are pre-computed.

In [14]:
import os

print(os.getcwd())

C:\Users\HOLE\DATA ANALYST\Python Projects\Revenue Leakage Detection & Profit Optimization\Python


In [ ]:
# sql to pandas

In [15]:
from sqlalchemy import create_engine

engine = create_engine(
    "mysql+pymysql://root:Rucha0915@localhost:3306/amazon_sales"
)

In [16]:
with engine.connect() as conn:
    print("Connection successful")

Connection successful


In [64]:
df.to_sql('Sales_Data', engine, if_exists='fail', index=False)

100000